data/summary/types/bgm.go

```diff
@@ -0,0 +1,244 @@
+
+	HasTimeInVeryLowRecords bool `json:"hasTimeInVeryLowRecords" bson:"hasTimeInVeryLowRecords"`
+	TimeInVeryLowRecords    *int `json:"timeInVeryLowRecords" bson:"timeInVeryLowRecords"`
+
+	HasTimeInHighPercent bool     `json:"hasTimeInHighPercent" bson:"hasTimeInHighPercent"`
+	TimeInHighPercent    *float64 `json:"timeInHighPercent" bson:"timeInHighPercent"`
+
+	HasTimeInHighRecords bool `json:"hasTimeInHighRecords" bson:"hasTimeInHighRecords"`
+	TimeInHighRecords    *int `json:"timeInHighRecords" bson:"timeInHighRecords"`
+
+	HasTimeInVeryHighPercent bool     `json:"hasTimeInVeryHighPercent" bson:"hasTimeInVeryHighPercent"`
+	TimeInVeryHighPercent    *float64 `json:"timeInVeryHighPercent" bson:"timeInVeryHighPercent"`
+
+	HasTimeInVeryHighRecords bool `json:"hasTimeInVeryHighRecords" bson:"hasTimeInVeryHighRecords"`
+	TimeInVeryHighRecords    *int `json:"timeInVeryHighRecords" bson:"timeInVeryHighRecords"`
+}
+
+type BGMPeriods map[string]BGMPeriod
+
+type BGMStats struct {
+	Periods    BGMPeriods                             `json:"periods" bson:"periods"`
+	Buckets    Buckets[BGMBucketData, *BGMBucketData] `json:"buckets" bson:"buckets"`
+	TotalHours int                                    `json:"totalHours" bson:"totalHours"`
+}
+
+func (*BGMStats) GetType() string {
+	return SummaryTypeBGM
+}
+
+func (*BGMStats) GetDeviceDataType() string {
+	return DeviceDataTypeBGM
+}
+
+func (s *BGMStats) Init() {
+	s.Buckets = make(Buckets[BGMBucketData, *BGMBucketData], 0)
+	s.Periods = make(map[string]BGMPeriod)
+	s.TotalHours = 0
+}
+
+func (s *BGMStats) GetBucketsLen() int {
+	return len(s.Buckets)
+}
+
+func (s *BGMStats) GetBucketDate(i int) time.Time {
+	return s.Buckets[i].Date
+}
+
+func (s *BGMStats) Update(userData any) error {
+	var err error
+	userDataTyped := userData.([]*glucoseDatum.Glucose)
+	err = AddData(&s.Buckets, userDataTyped)
+	if err != nil {
+		return err
+	}
+
+	s.CalculateSummary()
+
+	return nil
+}
+
+func (B *BGMBucketData) CalculateStats(r any, _ *time.Time) (bool, error) {
+	dataRecord, ok := r.(*glucoseDatum.Glucose)
+	if !ok {
+		return false, errors.New("BGM record for calculation is not compatible with Glucose type")
+	}
+	var normalizedValue float64
+
+	normalizedValue = *glucose.NormalizeValueForUnits(dataRecord.Value, pointer.FromString(summaryGlucoseUnits))
+
+	if normalizedValue < veryLowBloodGlucose {
+		B.VeryLowRecords++
+	} else if normalizedValue > veryHighBloodGlucose {
+		B.VeryHighRecords++
+	} else if normalizedValue < lowBloodGlucose {
+		B.LowRecords++
+	} else if normalizedValue > highBloodGlucose {
+		B.HighRecords++
+	} else {
+		B.TargetRecords++
+	}
+
+	B.TotalRecords++
+	B.TotalGlucose += normalizedValue
+
+	return false, nil
+}
+
+func (s *BGMStats) CalculateSummary() {
+	var totalStats = &BGMBucketData{}
+
+	// count backwards through hourly stats, stopping at 24, 24*7, 24*14, 24*30
+	// currently only supports day precision
+	stopPoints := []int{1, 7, 14, 30}
```

Comment:
Shouldn't stop points be defined one level up as a constant and shared between implementations?